File: src/main/java/com/github/zoewithabang/dao/BotMessageDao.java
package com.github.zoewithabang.dao;

import com.github.zoewithabang.model.BotMessage;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class BotMessageDao extends Dao<BotMessage, String>
{
    public BotMessageDao(Properties botProperties)
    {
        super(botProperties);
    }
    
    @Override
    public BotMessage get(Connection connection, String name) throws SQLException
    {
        String query = "SELECT * FROM bot_messages WHERE alias = ?;";
    
        try(PreparedStatement statement = connection.prepareStatement(query))
        {
            statement.setString(1, name);
        
            ResultSet resultSet = statement.executeQuery();
            BotMessage botMessage = null;
        
            if(resultSet.next())
            {
                String message = resultSet.getString("message");
                String description = resultSet.getString("description");
                botMessage = new BotMessage(name, message, description);
            }
        
            return botMessage;
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on getting BotMessage for name '{}'.", name, e);
            throw e;
        }
    }
    
    @Override
    public List<BotMessage> getAll(Connection connection) throws SQLException
    {
        String query = "SELECT * FROM bot_messages;";
    
        try(PreparedStatement statement = connection.prepareStatement(query))
        {
            ResultSet resultSet = statement.executeQuery();
            List<BotMessage> botMessageList = new ArrayList<>();
        
            while(resultSet.next())
            {
                String name = resultSet.getString("name");
                String message = resultSet.getString("message");
                String description = resultSet.getString("description");
                botMessageList.add(new BotMessage(name, message, description));
            }
        
            return botMessageList;
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on getting all BotMessages.", e);
            throw e;
        }
    }
    
    @Override
    public void store(Connection connection, BotMessage botMessage) throws SQLException
    {
        String query = "INSERT INTO bot_messages (`name`, `message`, `description`) VALUES (?, ?, ?);";
    
        try(PreparedStatement statement = connection.prepareStatement(query))
        {
            statement.setString(1, botMessage.getName());
            statement.setString(2, botMessage.getMessage());
            statement.setString(3, botMessage.getDescription());
        
            statement.executeUpdate();
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on storing new BotMessage '{}'.", botMessage, e);
            throw e;
        }
    }
    
    @Override
    public void update(Connection connection, BotMessage botMessage) throws SQLException
    {
        String query = "UPDATE bot_messages SET `message` = ?, `description` = ? WHERE `name` = ?;";
    
        try(PreparedStatement statement = connection.prepareStatement(query))
        {
            statement.setString(1, botMessage.getMessage());
            statement.setString(2, botMessage.getDescription());
            statement.setString(3, botMessage.getName());
        
            statement.executeUpdate();
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on updating BotMessage '{}'.", botMessage, e);
            throw e;
        }
    }
    
    @Override
    public void delete(Connection connection, BotMessage botMessage) throws SQLException
    {
        String query = "DELETE FROM bot_messages WHERE `name` = ?;";
    
        try(PreparedStatement statement = connection.prepareStatement(query))
        {
            statement.setString(1, botMessage.getName());
        
            statement.executeUpdate();
        }
        catch(SQLException e)
        {
            LOGGER.error("SQLException on deleting BotMessage '{}'.", botMessage, e);
            throw e;
        }
    }
}

File: src/main/kotlin/com/github/zoewithabang/dao/BotMessageDao.kt
package com.github.zoewithabang.dao

import com.github.zoewithabang.model.BotMessage
import org.apache.logging.log4j.LogManager

import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.util.ArrayList
import java.util.Properties

class BotMessageDao(botProperties: Properties) : Dao<BotMessage, String>(botProperties) {
    private val logger = LogManager.getLogger("BotMessageDao")

    @Override
    @Throws(SQLException::class)
    operator fun get(connection: Connection, name: String): BotMessage? {
        val query = "SELECT * FROM bot_messages WHERE `name` = ?;"

        try {
            connection.prepareStatement(query).use { statement ->
                statement.setString(1, name)

                val resultSet = statement.executeQuery()
                var botMessage: BotMessage? = null

                if (resultSet.next()) {
                    val message = resultSet.getString("message")
                    val description = resultSet.getString("description")
                    botMessage = BotMessage(name, message, description)
                }

                return botMessage
            }
        } catch (e: SQLException) {
            logger.error("SQLException on getting BotMessage for name '{}'.", name, e)

            throw e
        }

    }

    @Override
    @Throws(SQLException::class)
    fun getAll(connection: Connection): List<BotMessage> {
        val query = "SELECT * FROM bot_messages;"

        try {
            connection.prepareStatement(query).use { statement ->
                val resultSet = statement.executeQuery()
                val botMessageList = ArrayList<BotMessage>()

                while (resultSet.next()) {
                    val name = resultSet.getString("name")
                    val message = resultSet.getString("message")
                    val description = resultSet.getString("description")
                    botMessageList.add(BotMessage(name, message, description))
                }

                return botMessageList
            }
        } catch (e: SQLException) {
            logger.error("SQLException on getting all BotMessages.", e)

            throw e
        }

    }

    @Override
    @Throws(SQLException::class)
    fun store(connection: Connection, botMessage: BotMessage) {
        val query = "INSERT INTO bot_messages (`name`, `message`, `description`) VALUES (?, ?, ?);"

        try {
            connection.prepareStatement(query).use { statement ->
                statement.setString(1, botMessage.name)
                statement.setString(2, botMessage.message)
                statement.setString(3, botMessage.description)

                statement.executeUpdate()
            }
        } catch (e: SQLException) {
            logger.error("SQLException on storing new BotMessage '{}'.", botMessage, e)

            throw e
        }

    }

    @Override
    @Throws(SQLException::class)
    fun update(connection: Connection, botMessage: BotMessage) {
        val query = "UPDATE bot_messages SET `message` = ?, `description` = ? WHERE `name` = ?;"

        try {
            connection.prepareStatement(query).use { statement ->
                statement.setString(1, botMessage.message)
                statement.setString(2, botMessage.description)
                statement.setString(3, botMessage.name)

                statement.executeUpdate()
            }
        } catch (e: SQLException) {
            logger.error("SQLException on updating BotMessage '{}'.", botMessage, e)

            throw e
        }

    }

    @Override
    @Throws(SQLException::class)
    fun delete(connection: Connection, botMessage: BotMessage) {
        val query = "DELETE FROM bot_messages WHERE `name` = ?;"

        try {
            connection.prepareStatement(query).use { statement ->
                statement.setString(1, botMessage.name)

                statement.executeUpdate()
            }
        } catch (e: SQLException) {
            logger.error("SQLException on deleting BotMessage '{}'.", botMessage, e)

            throw e
        }

    }
}
